pyccel/ast/builtin_methods/list_methods.py

```diff
@@ -93,17 +94,19 @@ class ListPop(PyccelInternalFunction) :
     removes the item at the specified index. 
     The method also returns the removed item.
 
+    >>> [1, 2].pop()
+    2
+
     Parameters
     ----------
     list_variable : TypedAstNode
-        The name of the list.
+        The list object which the method is called from.
 
     index_element : TypedAstNode
         The current index value for the element to be popped.
     """
-    __slots__ = ('_dtype','_precision', '_index','_list_variable')
+    __slots__ = ('_dtype','_precision', '_index','_list_variable', '_rank')
     _attribute_nodes = ('_index','_list_variable')
-    _rank = 0
     _order = None
     _shape = None
```

Comment:
I think we should always have `len(self._shape) == self._rank` if `self._rank` is variable then shouldn't `_shape` also be?
https://github.com/pyccel/pyccel/blob/devel/developer_docs/ast_nodes.md#shape

Reply:
I agree
should i use `process_shape()` to get the new shape of the popped element, seeing that it's dependent on the rank

Reply:
`process_shape()` is designed to process the shape passed as an argument to a NumPy function and translate it to a shape that Pyccel can use. Is this the behaviour you are looking for:

https://github.com/pyccel/pyccel/blob/e371f27b97024cc1d5632029dc54a96e63023270/pyccel/ast/numpyext.py#L157-L158

Reply:
from the pattern that i noticed in these cases, something like this kinda made sense:
```python
a = [[[1, 2, 3], [4, 5, 6]]]
# shape(a) => (Literal(1), Literal(2), Literal(3))
b = a.pop() 
# b = [[1, 2, 3], [4, 5, 6]]
# shape(b) => (Literal(2),Literal(3))
c = b.pop() 
# c = [4, 5, 6]
# shape(c) = (Literal(3), )
d = c.pop()
# d = 6
# shape(d) = None
```
**Generalisation:**
```python
if len(list_variable.shape) == 1:
    self._shape = None
else:
    self_shape = tuple(list_variable.shape[1:])
```

Reply:
with the improved typing, something like:
```python
self._class_type = list_variable.dtype
```
?